File: backend/fastapi/app/services/extractor.py

import re
from ..models import CPRARequestDraft, CPRARequest, Requester, DateRange

def _parse_date(s: str) -> str:
    from dateutil import parser

    try:
        return parser.parse(s).date().isoformat()
    except Exception:
        return s


def extract_scope(notes: str) -> CPRARequestDraft:
    """Derive a draft CPRA request from a structured CPRA summary."""
    name = "Unknown Requester"
    email = "requester@example.com"
    m = re.search(r"Requester:\s*(.*?)\s*[\u2013\-]\s*(\S+)", notes)
    if m:
        name = m.group(1).strip()
        email = m.group(2).strip()
    received = "2025-01-01"
    m = re.search(r"Date Received:\s*([^\n]+)", notes)
    if m:
        received = _parse_date(m.group(1))
    matter = "No matter found"
    m = re.search(r"Matter:\s*([^\n]+)", notes)
    if m:
        matter = m.group(1).strip()
    range_start = None
    range_end = received
    m = re.search(r"Time Range:\s*([^\n]+)", notes)
    if m:
        parts = re.split(r"\s*[\u2013\-]\s*", m.group(1))
        if len(parts) == 2:
            range_start = _parse_date(parts[0])
            range_end = _parse_date(parts[1])
    record_types: list[str] = []
    record_text = ""
    m = re.search(r"Record Types:\s*([^\n]+)", notes)
    if m:
        record_text = m.group(1).strip()
        record_types = [s.strip() for s in re.split(r"[;,]", record_text) if s.strip()]
    custodians: list[str] = []
    m = re.search(r"Custodians:\s*([^\n]+)", notes)
    if m:
        custodians = [s.strip() for s in re.split(r"[;,]", m.group(1)) if s.strip()]
    pref = ""
    m = re.search(r"Preferred Format/Delivery:\s*([^\n]+)", notes)
    if m:
        pref = m.group(1).strip()
    draft = CPRARequest(
        requester=Requester(name=name, email=email),
        receivedDate=received,
        matter=matter,
        description=record_text,
        recordTypes=record_types,
        custodians=custodians,
        preferredFormatDelivery=pref,
        range=DateRange(start=range_start, end=range_end),
        departments=[],
    )
    return CPRARequestDraft(
        request=draft,
        confidences={"requester": 0.6, "receivedDate": 0.7, "matter": 0.6, "description": 0.6},
    )

File: backend/express/src/services/extractor.ts
import type { CPRARequest, CPRARequestDraft, Requester } from './types.js';

function parseDate(raw: string): string {
  const d = new Date(raw);
  return isNaN(d.getTime()) ? raw : d.toISOString().slice(0, 10);
}

/** Derive a draft CPRA request from a structured CPRA summary. */
export function extractScope(notes: string): CPRARequestDraft {
  let name = 'Unknown Requester';
  let email = 'requester@example.com';
  const rm = notes.match(/Requester:\s*(.*?)\s*[\u2013-]\s*(\S+)/);
  if (rm) {
    name = rm[1].trim();
    email = rm[2].trim();
  }
  let received = '2025-01-01';
  const dm = notes.match(/Date Received:\s*([^\n]+)/);
  if (dm) received = parseDate(dm[1]);
  let matter = 'No matter found';
  const sm = notes.match(/Matter:\s*([^\n]+)/);
  if (sm) matter = sm[1].trim();
  let rangeStart: string | undefined;
  let rangeEnd: string | undefined = received;
  const rmRange = notes.match(/Time Range:\s*([^\n]+)/);
  if (rmRange) {
    const parts = rmRange[1].split(/\s*[\u2013-]\s*/);
    if (parts.length === 2) {
      rangeStart = parseDate(parts[0]);
      rangeEnd = parseDate(parts[1]);
    }
  }
  let recordText = '';
  const rt = notes.match(/Record Types:\s*([^\n]+)/);
  const recordTypes = rt ? rt[1].split(/[;,]/).map(s => s.trim()).filter(Boolean) : [];
  if (rt) recordText = rt[1].trim();
  const cust = notes.match(/Custodians:\s*([^\n]+)/);
  const custodians = cust ? cust[1].split(/[;,]/).map(s => s.trim()).filter(Boolean) : [];
  const pf = notes.match(/Preferred Format\/Delivery:\s*([^\n]+)/);
  const preferred = pf ? pf[1].trim() : '';
  const requester: Requester = { name, email };
  const draft: CPRARequest = {
    requester,
    receivedDate: received,
    matter,
    description: recordText,
    recordTypes,
    custodians,
    preferredFormatDelivery: preferred,
    range: { start: rangeStart, end: rangeEnd },
    departments: [],
    extension: { apply: false, reasons: [] },
  };
  return {
    request: draft,
    confidences: { requester: 0.6, receivedDate: 0.7, matter: 0.6, description: 0.6 },
  };
}
